rosetta/transactions/intent.go

```diff
@@ -35,60 +36,46 @@ type Intent struct {
+	GasLimit       uint64
 }
 
-// CreateTransactionIntent creates a transaction Intent from two operations given as input.
+// DeriveIntent derives a transaction Intent from two operations given as input.
 // Specified operations should be symmetrical, a deposit and a withdrawal from two
 // different accounts. At the moment, the only fields taken into account are the
 // account IDs, amounts and type of operation.
-func (p *Parser) CreateTransactionIntent(operations []object.Operation) (*Intent, error) {
+func (p *Parser) DeriveIntent(operations []object.Operation) (*Intent, error) {
 
-	firstOpNegative := strings.HasPrefix(operations[0].Amount.Value, "-")
-	secondOpNegative := strings.HasPrefix(operations[1].Amount.Value, "-")
-
-	// Operations are invalid if both operations have negative amounts.
-	if firstOpNegative && secondOpNegative {
-
-		return nil, failure.InvalidIntent{
-			Description: failure.NewDescription("invalid operations - two deposits specified"),
-			Sender:      operations[0].AccountID.Address,
-			Receiver:    operations[1].AccountID.Address,
+	if len(operations) != 2 {
+		return nil, failure.InvalidOperations{
+			Description: failure.NewDescription("invalid number of operations"),
+			Count:       len(operations),
 		}
 	}
 
-	// Operations are also invalid if both operations have positive amounts.
-	if !firstOpNegative && !secondOpNegative {
-
-		return nil, failure.InvalidIntent{
-			Description: failure.NewDescription("invalid operations - two withdrawals specified"),
-			Sender:      operations[0].AccountID.Address,
-			Receiver:    operations[1].AccountID.Address,
-		}
-	}
+	// Sort the operations so that the send operation (negative amount) comes first.
+	sort.Slice(operations, func(i, j int) bool {
+		return operations[i].Amount.Value < operations[j].Amount.Value
+	})
 
-	// Assume the first operation is the one with the negative amount.
 	send := operations[0]
 	receive := operations[1]
 
-	// If that was not the case, switch the send and receive operations.
-	if !strings.HasPrefix(send.Amount.Value, "-") {
-		receive = operations[0]
-		send = operations[1]
-	}
-
-	// Validate the sender and the receiver account IDs.
-	err := p.validate.Account(send.AccountID)
-	if err != nil {
-		return nil, fmt.Errorf("invalid sender account: %w", err)
-	}
-	err = p.validate.Account(receive.AccountID)
-	if err != nil {
-		return nil, fmt.Errorf("invalid receiver account: %w", err)
+	// Verify send and receive values - send operation should be negative, receive operation should be positive.
+	if !strings.HasPrefix(send.Amount.Value, "-") ||
+		strings.HasPrefix(receive.Amount.Value, "-") {
+		return nil, failure.InvalidIntent{
+			Description: failure.NewDescription("invalid amounts for transfer",
+				failure.WithString("withdrawal_amount", send.Amount.Value),
+				failure.WithString("deposit_amount", receive.Amount.Value),
+			),
+			Sender:   send.AccountID.Address,
+			Receiver: receive.AccountID.Address,
+		}
 	}
```

Comment:
That check still seems overkill to me. Why not just parse both values and check that `send.Amount.Value` is `-receive.Amount.Value`? Doing string checks on numbers seems like a bad idea.